subprojects/wrapper-shared/src/main/java/org/gradle/wrapper/Install.java

```diff
@@ -113,8 +94,85 @@ public File call() throws Exception {
         });
     }
 
-    private String calculateSha256Sum(File file)
-        throws Exception {
+    private void fetchDistribution(URI safeDistributionUrl, File localZipFile, URI distributionUrl, File distDir, WrapperConfiguration configuration) throws Exception {
+        String distributionSha256Sum = configuration.getDistributionSha256Sum();
+        boolean failed = false;
+        int retries = RETRIES;
+        do {
+            try {
+                boolean needsDownload = !localZipFile.isFile() || failed;
+                if (needsDownload) {
+                    forceFetch(safeDistributionUrl, localZipFile, distributionUrl);
+                }
+
+                deleteLocalTopLevelDirs(distDir);
+
+                verifyDownloadChecksum(configuration.getDistribution().toString(), localZipFile, distributionSha256Sum);
+
+                unzipLocal(localZipFile, distDir);
+                failed = false;
+            } catch (ZipException e) {
```

Comment:
I would say not to rely on such a specific exception here, but use a more generic one, like `IOException`. The specific exception makes the code brittle. For example we have recently changed the compression library we use, when we do it again a completely different exception might get thrown. Btw. is there a test that would catch this?

Reply:
we won't change it here though?
I assume the wrapper code should be as independent of 3rd party libraries as possible, hence we will stick to the java basics, right?
And I don't think we have a good strategy to recover from a generic IOException, at least the old implementation would just let it through.

Reply:
Maybe we will change it, but we should have a test covering it anyways. I still think `ZipException` is too specific. In my experience the compression libraries aren't consistent enough. At least do some testing with:
* empty zip file
* tar file with zip extension
* completely jumbled binary file 
* test file